### PR TITLE
Basic services implementation

### PR DESCRIPTION
**Pull request by mosquito**

### CODE REVIEW

File: aiomisc/utils.py [hunk 1]
```diff
@@ -46,9 +47,9 @@ def bind_socket(*args, address: str, port: int, options=(),
     sock_addr = address, port
 
     if sock.family == socket.AF_INET6:
-        log.info('Listening tcp://[%s]:%s' % sock_addr)
+        log.info('Listening %s://[%s]:%s', proto_name, *sock_addr)
```
Comment (leenr) [inline aiomisc/utils.py]: If address is IPv6 one, sock_addr will be a 4-tuple.
https://github.com/mosquito/carbon-proxy/pull/5

File: tests/test_entrypoint.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+def test_tcp_server(unused_tcp_port):
+    class TestService(TCPServer):
+        DATA = []
+
+        async def handle_client(self, reader: asyncio.StreamReader,
+                                writer: asyncio.StreamWriter):
+            self.DATA.append(await reader.readline())
+            writer.close()
+
+    service = TestService('127.0.0.1', unused_tcp_port)
```
Comment (leenr) [inline tests/test_entrypoint.py]: `pytest.mark.parametrize` IPv4 and IPv6 localhost?

File: aiomisc/utils.py [hunk 3]
```diff
@@ -22,11 +22,12 @@ def chunk_list(iterable: Iterable[Any], size: int):
         item = list(itertools.islice(iterable, size))
 
 
-OptionsType = List[Tuple[int, int, int]]
+OptionsType = Tuple[Tuple[int, int, int], ...]
```
Comment (leenr) [inline aiomisc/utils.py]: Maybe better to use `Iterable` here?

File: aiomisc/service.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            )
+        elif not isinstance(sock, socket.socket):
+            raise ValueError('sock must be socket instance')
+        else:
+            self.socket = sock
+
+        self.server = None
+        super().__init__()
+
+    def handle_datagram(self, data: bytes, addr):
```
Comment (leenr) [inline aiomisc/service.py]: Type annotation for `addr`?

File: aiomisc/service.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+        def connection_made(self, transport: asyncio.DatagramTransport):
+            self.transport = transport
+            self.loop = asyncio.get_event_loop()
+
+        def datagram_received(self, data: bytes, addr):
+            self.loop.create_task(self.handler(data, addr))
+
+    def __init__(self, address: str=None, port: int=None,
+                 options=(), sock=None):
```
Comment (leenr) [inline aiomisc/service.py]: Type annotation for `options` and `sock`.

File: tests/test_entrypoint.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        DATA = []
+
+        async def handle_datagram(self, data: bytes, addr):
+            self.DATA.append(data)
+
+    service = TestService('127.0.0.1', unused_tcp_port)
+
+    @threaded
+    def writer():
+        sock = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
```
Comment (leenr) [inline tests/test_entrypoint.py]: `socket.AF_INET6`, but IPv4 address below.

File: aiomisc/utils.py [hunk 7]
```diff
@@ -22,11 +22,12 @@ def chunk_list(iterable: Iterable[Any], size: int):
...
 
 
-def bind_socket(*args, address: str, port: int, options=(),
-                reuse_addr=True, reuse_port=False):
+def bind_socket(*args, address: str, port: int, options: OptionsType = (),
+                reuse_addr: bool = True, reuse_port: bool = False,
+                proto_name: str = 'tcp'):
```
Comment (leenr) [inline aiomisc/utils.py]: Maybe it's better to define enum for `proto_name`?

Comment (mosquito) [reply aiomisc/utils.py]: No. The proto name is user defined.

**Pull request merged**